src/analysis/inner_product_space/pi_L2.lean

```diff
@@ -364,23 +460,148 @@ by simp [complex.isometry_of_orthonormal, (dec_trivial : (finset.univ : finset (
+  [decidable_eq E]
+  (hV' : @orthogonal_family 𝕜 _ _ _ _ (λ i, V i) _ (λ i, (V i).subtypeₗᵢ)) :
+  (Σ i, orthonormal_basis_index 𝕜 (V i)) ≃ fin n :=
+let b := hV.collected_orthonormal_basis hV' (λ i, (std_orthonormal_basis 𝕜 (V i))) in
+fintype.equiv_fin_of_card_eq $ (finite_dimensional.finrank_eq_card_basis b.to_basis).symm.trans hn
+
+/-- An `n`-dimensional `inner_product_space` equipped with a decomposition as an internal direct
+sum has an orthonormal basis indexed by `fin n` and subordinate to that direct sum. -/
+@[irreducible] def direct_sum.submodule_is_internal.subordinate_orthonormal_basis
+  [decidable_eq E]
+  (hV' : @orthogonal_family 𝕜 _ _ _ _ (λ i, V i) _ (λ i, (V i).subtypeₗᵢ)) :
+  orthonormal_basis (fin n) 𝕜 E :=
+((hV.collected_orthonormal_basis hV' (λ i, (std_orthonormal_basis 𝕜 (V i)))).reindex
+  (hV.sigma_orthonormal_basis_index_equiv hn hV'))
+
+/-- An `n`-dimensional `inner_product_space` equipped with a decomposition as an internal direct
+sum has an orthonormal basis indexed by `fin n` and subordinate to that direct sum. This function
+provides the mapping by which it is subordinate. -/
+def direct_sum.submodule_is_internal.subordinate_orthonormal_basis_index [decidable_eq E]
+  (a : fin n) (hV' : @orthogonal_family 𝕜 _ _ _ _ (λ i, V i) _ (λ i, (V i).subtypeₗᵢ)) : ι :=
+((hV.sigma_orthonormal_basis_index_equiv hn hV').symm a).1
+
+/-- The basis constructed in `orthogonal_family.subordinate_orthonormal_basis` is subordinate to
+the `orthogonal_family` in question. -/
+lemma direct_sum.submodule_is_internal.subordinate_orthonormal_basis_subordinate [decidable_eq E]
+  (a : fin n) (hV' : @orthogonal_family 𝕜 _ _ _ _ (λ i, V i) _ (λ i, (V i).subtypeₗᵢ)) :
+  (hV.subordinate_orthonormal_basis hn hV' a) ∈
+  V (hV.subordinate_orthonormal_basis_index hn a hV') :=
+by simpa only [direct_sum.submodule_is_internal.subordinate_orthonormal_basis,
+  orthonormal_basis.coe_reindex]
+  using hV.collected_orthonormal_basis_mem hV' (λ i, (std_orthonormal_basis 𝕜 (V i)))
+    ((hV.sigma_orthonormal_basis_index_equiv hn hV').symm a)
+
+attribute [irreducible] direct_sum.submodule_is_internal.subordinate_orthonormal_basis_index
+
+end subordinate_orthonormal_basis
+
+end finite_dimensional
 
 local attribute [instance] fact_finite_dimensional_of_finrank_eq_succ
 
 /-- Given a natural number `n` one less than the `finrank` of a finite-dimensional inner product
 space, there exists an isometry from the orthogonal complement of a nonzero singleton to
 `euclidean_space 𝕜 (fin n)`. -/
 def linear_isometry_equiv.from_orthogonal_span_singleton
-  (n : ℕ) [fact (finrank 𝕜 E = n + 1)] {v : E} (hv : v ≠ 0) :
-  (𝕜 ∙ v)ᗮ ≃ₗᵢ[𝕜] (euclidean_space 𝕜 (fin n)) :=
-linear_isometry_equiv.of_inner_product_space (finrank_orthogonal_span_singleton hv)
+  [decidable_eq E] (n : ℕ) [fact (finrank 𝕜 E = n + 1)] {v : E} (hv : v ≠ 0) :
+  orthonormal_basis (fin n) 𝕜 (𝕜 ∙ v)ᗮ :=
+(fin_std_orthonormal_basis (finrank_orthogonal_span_singleton hv))
```

Comment:
Let's rename this now that the type has changed (thanks!) -- say to `orthonormal_basis.orthogonal_span_singleton`